File: examples/nodejs/demo-server.js
/**
 * FPScanner Demo Server (Node.js)
 * 
 * This server demonstrates how to receive and decrypt fingerprints
 * collected by fpscanner on the client side.
 */

const http = require('http');
const fs = require('fs');
const path = require('path');

const PORT = 3000;

// The encryption key - must match the key used when building fpscanner
// In production, this should come from environment variables
const ENCRYPTION_KEY = process.env.FINGERPRINT_KEY || 'dev-key';

/**
 * Decrypt an XOR-encrypted, base64-encoded string
 */
function decryptString(ciphertext, key) {
    const binaryString = Buffer.from(ciphertext, 'base64').toString('binary');
    const encrypted = new Uint8Array(binaryString.length);
    for (let i = 0; i < binaryString.length; i++) {
        encrypted[i] = binaryString.charCodeAt(i);
    }

    const keyBytes = Buffer.from(key, 'utf8');
    const decrypted = new Uint8Array(encrypted.length);

    for (let i = 0; i < encrypted.length; i++) {
        decrypted[i] = encrypted[i] ^ keyBytes[i % keyBytes.length];
    }

    return Buffer.from(decrypted).toString('utf8');
}

/**
 * Decrypt and parse a fingerprint payload
 */
function decryptFingerprint(encryptedFingerprint) {
    const decryptedJson = decryptString(encryptedFingerprint, ENCRYPTION_KEY);
    let parsed = JSON.parse(decryptedJson);
    // Handle double-JSON-encoding (string containing JSON)
    if (typeof parsed === 'string') {
        parsed = JSON.parse(parsed);
    }
    return parsed;
}

// MIME types for serving static files
const MIME_TYPES = {
    '.html': 'text/html',
    '.js': 'application/javascript',
    '.css': 'text/css',
    '.json': 'application/json',
};

const server = http.createServer(async (req, res) => {
    // Handle fingerprint submission
    if (req.method === 'POST' && req.url === '/api/fingerprint') {
        let body = '';
        req.on('data', chunk => body += chunk);
        req.on('end', () => {
            try {
                const { fingerprint: encryptedFingerprint } = JSON.parse(body);
                
                // Decrypt the fingerprint
                const fingerprint = decryptFingerprint(encryptedFingerprint);
                
                // Log the decrypted fingerprint
                console.log('\n' + '='.repeat(60));
                console.log('📥 Received fingerprint from client');
                console.log('='.repeat(60));
                console.log('\n🔓 Decrypted fingerprint:');
                console.log(JSON.stringify(fingerprint, null, 2));
                console.log('\n📊 Summary:');
                console.log(`   FSID: ${fingerprint.fsid}`);
                console.log(`   Platform: ${fingerprint.signals.device.platform}`);
                console.log(`   User Agent: ${fingerprint.signals.browser.userAgent.substring(0, 50)}...`);
                console.log(`   CPU Count: ${fingerprint.signals.device.cpuCount}`);
                console.log(`   Memory: ${fingerprint.signals.device.memory} GB`);
                console.log(`   Screen: ${fingerprint.signals.device.screenResolution.width}x${fingerprint.signals.device.screenResolution.height}`);
                console.log(`   Bot Detection: ${fingerprint.fastBotDetection ? '⚠️  SUSPICIOUS' : '✓ OK'}`);
                console.log('='.repeat(60) + '\n');
                
                // Send response with full fingerprint
                res.writeHead(200, { 'Content-Type': 'application/json' });
                res.end(JSON.stringify({
                    success: true,
                    fingerprint: fingerprint
                }));
            } catch (error) {
                console.error('❌ Error processing fingerprint:', error.message);
                res.writeHead(400, { 'Content-Type': 'application/json' });
                res.end(JSON.stringify({ success: false, error: error.message }));
            }
        });
        return;
    }
    
    // Serve static files
    let filePath;
    if (req.url === '/' || req.url === '/index.html') {
        filePath = path.join(__dirname, 'index.html');
    } else {
        // Serve files from the fpscanner root (for dist folder access)
        filePath = path.join(__dirname, '../..', req.url);
    }
    
    const ext = path.extname(filePath);
    const contentType = MIME_TYPES[ext] || 'text/plain';
    
    fs.readFile(filePath, (err, content) => {
        if (err) {
            res.writeHead(404);
            res.end(`Not found: ${req.url}`);
            return;
        }
        res.writeHead(200, { 'Content-Type': contentType });
        res.end(content);
    });
});

server.listen(PORT, () => {
    console.log(`
╔════════════════════════════════════════════════════════╗
║          FPScanner Demo Server (Node.js)               ║
╠════════════════════════════════════════════════════════╣
║  Server running at: http://localhost:${PORT}              ║
║  Open this URL in your browser to test                 ║
╚════════════════════════════════════════════════════════╝
`);
});

File: examples/python/demo-server.py
#!/usr/bin/env python3
"""
FPScanner Demo Server (Python)

This server demonstrates how to receive and decrypt fingerprints
collected by fpscanner on the client side.
"""

import base64
import json
import os
from http.server import HTTPServer, SimpleHTTPRequestHandler
from urllib.parse import urlparse

PORT = 3010

# The encryption key - must match the key used when building fpscanner
# In production, this should come from environment variables
ENCRYPTION_KEY = os.environ.get('FINGERPRINT_KEY', 'dev-key')


def xor_decrypt(ciphertext_b64: str, key: str) -> str:
    """
    Decrypt an XOR-encrypted, base64-encoded string.
    
    Args:
        ciphertext_b64: Base64-encoded encrypted data
        key: The encryption key (must match the key used for encryption)
    
    Returns:
        Decrypted string
    """
    # Decode from base64
    encrypted = base64.b64decode(ciphertext_b64)
    key_bytes = key.encode('utf-8')
    
    # XOR decrypt
    decrypted = bytearray(len(encrypted))
    for i in range(len(encrypted)):
        decrypted[i] = encrypted[i] ^ key_bytes[i % len(key_bytes)]
    
    return decrypted.decode('utf-8')


def decrypt_fingerprint(encrypted_fingerprint: str) -> dict:
    """
    Decrypt and parse a fingerprint payload.
    
    Args:
        encrypted_fingerprint: The encrypted fingerprint from the client
    
    Returns:
        Parsed fingerprint dictionary
    """
    decrypted_json = xor_decrypt(encrypted_fingerprint, ENCRYPTION_KEY)
    parsed = json.loads(decrypted_json)
    # Handle double-JSON-encoding (string containing JSON)
    if isinstance(parsed, str):
        parsed = json.loads(parsed)
    return parsed


class FingerprintHandler(SimpleHTTPRequestHandler):
    """HTTP request handler for the fingerprint demo."""
    
    def __init__(self, *args, **kwargs):
        # Set the directory to serve static files from
        super().__init__(*args, directory=os.path.dirname(os.path.abspath(__file__)), **kwargs)
    
    def do_POST(self):
        """Handle POST requests (fingerprint submission)."""
        if self.path == '/api/fingerprint':
            try:
                # Read the request body
                content_length = int(self.headers['Content-Length'])
                body = self.rfile.read(content_length).decode('utf-8')
                data = json.loads(body)
                
                encrypted_fingerprint = data.get('fingerprint')
                if not encrypted_fingerprint:
                    raise ValueError('No fingerprint provided')
                
                # Decrypt the fingerprint
                fingerprint = decrypt_fingerprint(encrypted_fingerprint)
                
                # Log the decrypted fingerprint
                print('\n' + '=' * 60)
                print('📥 Received fingerprint from client')
                print('=' * 60)
                print('\n🔓 Decrypted fingerprint:')
                print(json.dumps(fingerprint, indent=2))
                print('\n📊 Summary:')
                print(f"   FSID: {fingerprint['fsid']}")
                print(f"   Platform: {fingerprint['signals']['device']['platform']}")
                print(f"   User Agent: {fingerprint['signals']['browser']['userAgent'][:50]}...")
                print(f"   CPU Count: {fingerprint['signals']['device']['cpuCount']}")
                print(f"   Memory: {fingerprint['signals']['device']['memory']} GB")
                screen = fingerprint['signals']['device']['screenResolution']
                print(f"   Screen: {screen['width']}x{screen['height']}")
                bot_status = '⚠️  SUSPICIOUS' if fingerprint['fastBotDetection'] else '✓ OK'
                print(f'   Bot Detection: {bot_status}')
                print('=' * 60 + '\n')
                
                # Send response with full fingerprint
                response = {
                    'success': True,
                    'fingerprint': fingerprint
                }
                self._send_json(200, response)
                
            except Exception as e:
                print(f'❌ Error processing fingerprint: {e}')
                self._send_json(400, {'success': False, 'error': str(e)})
        else:
            self.send_error(404, 'Not Found')
    
    def do_GET(self):
        """Handle GET requests (serve static files)."""
        parsed = urlparse(self.path)
        path = parsed.path
        
        # Serve index.html for root
        if path == '/' or path == '/index.html':
            self.path = '/index.html'
            return super().do_GET()
        
        # Serve files from fpscanner root (for dist folder access)
        if path.startswith('/dist/'):
            # Construct path relative to fpscanner root
            file_path = os.path.join(
                os.path.dirname(os.path.abspath(__file__)),
                '../..',
                path.lstrip('/')
            )
            if os.path.exists(file_path):
                self.send_response(200)
                if path.endswith('.js'):
                    self.send_header('Content-Type', 'application/javascript')
                else:
                    self.send_header('Content-Type', 'application/octet-stream')
                self.end_headers()
                with open(file_path, 'rb') as f:
                    self.wfile.write(f.read())
                return
        
        return super().do_GET()
    
    def _send_json(self, status: int, data: dict):
        """Send a JSON response."""
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        self.wfile.write(json.dumps(data).encode('utf-8'))
    
    def log_message(self, format, *args):
        """Suppress default logging for cleaner output."""
        if 'POST /api/fingerprint' not in format % args:
            # Only log non-fingerprint requests
            pass


def main():
    server = HTTPServer(('', PORT), FingerprintHandler)
    print(f'''
╔════════════════════════════════════════════════════════╗
║          FPScanner Demo Server (Python)                ║
╠════════════════════════════════════════════════════════╣
║  Server running at: http://localhost:{PORT}              ║
║  Open this URL in your browser to test                 ║
╚════════════════════════════════════════════════════════╝
''')
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print('\nServer stopped.')


if __name__ == '__main__':
    main()
